poc/observe.py

```diff
@@ -0,0 +1,701 @@
+        depends on the type of events. For example, a change to a simple
+        trait will emit CTraitObserverEvent. A mutation to a list
+        will emit ListObserverEvent.
+    path : ListenerPath
+        Path for listening to extended traits.
+    remove : boolean
+        Whether to remove the observer.
+    dispatch : str
+        A string indicating the thread on which notifications should be
+        run.
+    """
+    if remove:
+        remove_notifiers(
+            path=path,
+            object=object,
+            callback=callback,
+            target=object,
+            dispatcher=DISPATCHERS[dispatch],
+        )
+    else:
+        add_notifiers(
+            path=path,
+            object=object,
+            callback=callback,
+            target=object,
+            dispatcher=DISPATCHERS[dispatch],
+        )
+
+
+def add_notifiers(object, callback, path, target, dispatcher):
+    """ Add notifiers for a ListenerPath
+
+    Parameters
+    ----------
+    object : INotifiableObject
+        An object that implements `_notifiers` for returning a list for
+        adding notifiers
+    callback : callable(event)
+        A callable conforming to the notifier signature. ``event`` is
+        an object representing the change to notify about. Its type
+        and content depends on the change.
+    path : ListenerPath
+        Path for listening to extended traits.
+    target : Any
+        Sets the context for the (actual) callback. In practice, it is
+        the HasTrait instance on which the listener path is defined and
+        will be seen by users as the "owner" of the notifier.
+        Strictly speaking, this object sets the context for the notifier
+        and does not have to be a notifiable object.
+    dispatcher : callable(callable, args, kwargs)
+        Callable for dispatching the callback, i.e. dispatching
+        callback on a different thread.
+    """
+
+    listener = path.node
+    for this_target in listener.iter_this_targets(object):
+        if listener.notify:
+            notifier = listener.create_user_notifier(
+                object=this_target,
+                callback=callback,
+                target=target,
+                dispatcher=dispatcher,
+            )
+            notifier.add_to(object=this_target)
+
+        for next_path in path.nexts:
+
+            change_notifier = listener.get_change_notifier(
+                callback=callback,
+                path=next_path,
+                target=target,
+                dispatcher=dispatcher,
+            )
+            change_notifier.add_to(object=this_target)
+
+            for next_target in listener.iter_next_targets(object):
+                add_notifiers(
+                    object=next_target,
+                    callback=callback,
+                    path=next_path,
+                    target=target,
+                    dispatcher=dispatcher,
+                )
+    if type(listener) is not TraitAddedListener:
```

Comment:
Is there a way that this could be absorbed into the `BaseListener` "interface"? This explicit type check seems like a form of OCP violation.

Reply:
Yes!

Reply:
Note: This check is to prevent recursion, because it calls `add_notifiers` again with `TraitAddedListener` as the first listener.

Reply:
Done: https://github.com/enthought/traits/pull/942/commits/7e87fb35476180a042b9b525c42ef42ffb0490ea

Now the recursion is avoided by the `TraitAddedListener.get_static_paths` not returning any paths that have `TraitAddedListener` as root (it returns an empty list at the moment).